### DISCUSSION TITLE
GSOC 2022: Interested in project "Windows support for native SQLite"

### DISCUSSION THREAD

**Discussion by rishits321** [robolectric/robolectric#7155] [General]
Hello developers,

I am Rishikesh Suryavanshi and I am currently pursuing a bachelor's in Information Technology at Pune Institute of Computer Technology.
I am well versed with java and c++ programming languages and have good knowledge of working on native SQLite.
I have a firm grasp of data structures and algorithms and object-oriented programming. I am also familiar with windows programming. This project falls well in my current capabilities and I will be very excited to work on this.

Can you give me more details as to why Robolectric native runtime does not run on windows? 
I would like to connect with mentors for this project to guide me, I also want to get familiar with the codebase of Robolectric so any important guidelines or tips will be much appreciated.
I look forward to hearing from you and further working on this project.

**Comment by hoisie** [comment]
@rishits321 thanks for reaching out about this project. I would be happy to get you started on it.

At a high-level, currently, in Robolectric on Windows, the `android.database` package is primarily backed by [ShadowLegacySQLiteConnection](https://github.com/robolectric/robolectric/blob/master/shadows/framework/src/main/java/org/robolectric/shadows/ShadowLegacySQLiteConnection.java). This has many downsides -- mainly that it is using an obsolete sqlite4java library that does not support Mac M1. ShadowLegacySQLiteConnection also does not support Android extensions (e.g. phone_numbers_equal, localized collators), as well as missing a bunch of recent Android sqlite features (e.g. setting custom scalar functions).

In response to these shortcomings of sqlite4java, we added a new native SQLite mode that builds Android's version of sqlite + extensions in Robolectric sources (see https://github.com/robolectric/robolectric/tree/master/nativeruntime/external).

This new mode works great for Mac and Linux, but it does not work using Windows.

The best first step would be to get the build environment set up. You will need to install [msys2](https://www.msys2.org/) (MINGW64 environment) , and then use `pacman` to install git, ninja-build, and cmake. You may also need to install Adoptium OpenJDK 11 for JAVA_HOME.

Then try running:
```
 export JAVA_HOME="/c/Program Files/Eclipse Adoptium/jdk-11.0.13.8-hotspot/"
 ./gradlew :robolectric:test --tests='*SQLite*' -Drobolectric.sqliteMode=NATIVE
 ```

And you should see errors such as:

```
 android.database.sqlite.SQLiteException: not an error (code 0 SQLITE_OK): Failed to clear the cursor window, status=-40
```

This is the crux of the problem :)

**Reply by hoisie** [reply]
My suggestion would be to checkout the code in an msys2(MINGW64 environment), and see how far you can get to building. Unfortunately there is not a lot of Windows documentation about this at the moment, but we can work on improve on that. Please describe any issues you encounter and we can help unblock you.

**Reply by rishits321** [reply]
Thank you for replying. So I understood that the new SQLite mode does not work with Windows. I will follow the steps mentioned by you and get back to you soon if I face any problems.
